---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# mixturesnpcutoff

<!-- badges: start -->

<!-- badges: end -->

This package implements a mixture distribution approach to estimating mutation rates and SNP thresholds for paired metagenomic data. It also provides some plotting functions and simulation functions.

## Installation

You can install the development version of mixturesnpcutoff from [GitHub](https://github.com/) with:

``` r
install.packages("pak")
pak::pak("ZunairKhm/mixturesnpcutoff")
```

## Example

This shows the basic use case using data produced from the simulation functions.

```{r example}
library(mixturesnpcutoff)
## basic example code
x <- simulate_mixsnp_data(0.00179*4, 0.8, 500)
y <- simulate_unrelsnp_data()

z<- mixture_snp_cutoff(x$snp_dist, y, x$time_dist)
```

mixture_snp_cutoff will produce a mutation rate in SNPs/day or SNPs/day/million basepairs if sites considered is provided. It will also produce a cutoff that considers a max time (default 2 years) for SNP distances that can be considered related/transmitted or unrelated.
