---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# MxSure

<!-- badges: start -->

[![R-CMD-check](https://github.com/zunairkhm/mxsure/workflows/R-CMD-check/badge.svg)](https://github.com/zunairkhm/mxsure/actions)

<picture>
  <source media="(prefers-color-scheme: dark)" srcset="man/figures/logo_dark.svg">
  <source media="(prefers-color-scheme: light)" srcset="man/figures/logo_light.svg">
  <img alt="MxSure logo" src="man/figures/logo_light.svg">
</picture>


<!-- [![DOI](https://zenodo.org/badge/XXXX.svg)](https://zenodo.org/badge/latestdoi/XXXX) -->

<!-- badges: end -->

This package implements a mixture distribution approach to estimating substitution rates and SNP thresholds from pairwise SNP comparison data, in a longitudinal/within-host or transmission setting. It also provides some plotting and simulation functions.

## Installation

You can install the development version of mxsure from [GitHub](https://github.com/) with:

```{r, message=FALSE}
if(!require(remotes)) install.packages("remotes")
remotes::install_github("ZunairKhm/mxsure", quiet=T)
library(mxsure)
```

## Loading in data

MxSure can input data from inStrain compare output, TRACS distance output, or from a distance matrix. Either provide an R object or filepath to the respective function and the output will be an R object ready for inputting into MxSure. These examples only show loading in one dataset for each method however a mixed and distant dataset is required for MxSure.

### TRACS distance output

```{r}
#loading example data provided with package
tracs_example_path <- system.file("extdata", "example_tracs_output.csv", package = "mxsure")
sample_dates <- readr::read_csv(system.file("extdata", "example_sampling_dates.csv", package = "mxsure"))
head(sample_dates)

#using function to ready data for input into mxsure
mxsure_input_tracs <-  mxsure_input_tracs(file_path=tracs_example_path, dates=sample_dates, dates_sample_col = "sample_id", dates_date_col = "date")
head(mxsure_input_tracs)

```

### inStrain compare output

```{r}
instrain_example_path <- system.file("extdata", "example_instrain_output.csv", package = "mxsure")
sample_dates <- readr::read_csv(system.file("extdata", "example_sampling_dates.csv", package = "mxsure"))
mxsure_input_instrain <-  mxsure_input_instrain(file_path=instrain_example_path, dates=sample_dates, dates_sample_col = "sample_id", dates_date_col = "date")
head(mxsure_input_instrain)
```

### distance matrix

note: this function does not allow for 'sites considered' or equivalent; this will need to be inputed manually

```{r}
distmat_example_path <- system.file("extdata", "example_distance_matrix.csv", package = "mxsure")
sample_dates <- readr::read_csv(system.file("extdata", "example_sampling_dates.csv", package = "mxsure"))
mxsure_input_distmat <-  mxsure_input_distmatrix(file_path=distmat_example_path, dates=sample_dates, dates_sample_col = "sample_id", dates_date_col = "date")
head(mxsure_input_distmat)
```

## Basic Operation

MxSure in an R package designed to estimate substitution rates and infer SNP thresholds from longitudinally collected pairwise SNP comparison data. This data is likely to contain a mixture of related pairs (from longitudinal samples of the same microbe showing evolution over time) and unrelated pairs (co-captured samples from the same strain/species).

To demonstrate its use we can first simulate some example data using the `simulate_mixsnp_data()` command. Here we simulate 100 data points (each representing a pairwise SNP comparison between two sequences) with a true substitution rate of 5 SNPs/year with each data point having an 80% chance to be a related data point. By default sampling time differences for each of these data points are between 0 and 1 year.

```{r setup}
library(mxsure)
set.seed(123)
mixed_data <- simulate_mixsnp_data(lambda=5, k=0.8, n=100)
head(mixed_data, 5)
```

MxSure's methodology is to fit a mixture distribution, which describes the related and unrelated SNP distances in separate models, to this dataset. However, to do this we need to fit the unrelated SNP distance model (a right truncated negative binomial distribution) on another dataset of pairwise SNP comparisons that has no related pairs. This could be comparisons between different individuals and ideally individuals that are geographically separated to exclude transmission events. Here we again simulate the data using the same true substitution rate but with related proportion, k, set to 0.

```{r}
distant_data <- simulate_mixsnp_data(lambda=5, k=0, n=1000)
head(distant_data, 5)
```

We can then use `mxsure_estimate()` to estimate the substitution rate, estimate the related proportion, and infer a SNP threshold.

```{r}
result <- mxsure_estimate(
  mixed_snp_dist = mixed_data$snp_dist,
  unrelated_snp_dist = distant_data$snp_dist,
  mixed_time_dist = mixed_data$time_dist
)
result
```

Here lambda is the estimated substitution rate and k is the estimated related proportion. Intercept is a time independent parameter that describes the expected SNPs at time 0, which can be used to model sequencing error. Estimated fp displayed the proportion of the distant dataset that has a SNP distance below the inferred SNP threshold. nb_size and nb_mu are the parameters of the unrelated SNP distance model.

As in this example, if sites considered are not supplied for each comparison in the mixed dataset the estimated rate will have units SNPs/year/genome. We can display what this by simulating some amount of sites.

```{r}
mixed_data_sites <- mixed_data
mixed_data_sites$sites <- runif(100, 0.5e6, 1.5e6)

result_sites <- mxsure_estimate(
  mixed_snp_dist = mixed_data_sites$snp_dist,
  unrelated_snp_dist = distant_data$snp_dist,
  mixed_time_dist = mixed_data_sites$time_dist,
  mixed_sites = mixed_data_sites$sites
)
result_sites
```

## Youden and Threshold Range

For threshold assessment it may be useful to compare the MxSure threshold with what would be inferred from the Youden method.

```{r}
result <- mxsure_estimate(
  mixed_snp_dist = mixed_data$snp_dist,
  unrelated_snp_dist = distant_data$snp_dist,
  mixed_time_dist = mixed_data$time_dist,
  youden = TRUE
)
result$youden
```

It may also be useful to examine how a different threshold time affects the MxSure threshold.

```{r}
result <- mxsure_estimate(
  mixed_snp_dist = mixed_data$snp_dist,
  unrelated_snp_dist = distant_data$snp_dist,
  mixed_time_dist = mixed_data$time_dist,
  threshold_range = TRUE
)
result$threshold_range
```

## Confidence Intervals

Confidence intervals for substitution rate estimation are produced via bootstrapping. This can be computed faster with multi-threading using the future package. 20 bootstraps is very likely too few, you would likely need at least 200 but ideally 500 or 1000.

```{r}
future::plan("future::multisession", workers=2)
ci <- mxsure_ci( mixed_snp_dist = mixed_data$snp_dist,
  unrelated_snp_dist = distant_data$snp_dist,
  mixed_time_dist = mixed_data$time_dist,
  bootstraps=20
  )
future::plan("future::sequential")
ci$confidence_intervals
```

## Time Randomisation Test

This is a test that validates there is temporal signal in the underlying data. It achieves this by randomly permuting the time column in the mixed dataset and produces estimates and 95% CIs on this time randomised dataset. This is repeated on multiple time randomisation datasets, 10 is recommended, and if more than 5% of these bootstrap samples are above or equal to the original estimate this test is failed. Estimates should not be considered to be valid if this test fails. The output of this function is a graph showing this result and an outcome table.

```{r}
future::plan("future::multisession", workers=2)
timerand <- mxsure_timerandtest( mixed_snp_dist = mixed_data$snp_dist,
  unrelated_snp_dist = distant_data$snp_dist,
  mixed_time_dist = mixed_data$time_dist,
  bootstraps=20,
  permutations=2,
  original_result = result$results,
  ci_data = ci
  )
future::plan("future::sequential")
timerand$plot
timerand$outcome
```

## Likelihood ratios, and SNP over Time plot

For each data point in the mixed dataset, after fitting, we can assess the likelihood that any is related or unrelated. This is achieved using the `mxsure_likelihood()` function which reports likelihoods, log likelihoods, and likelihood ratios.

```{r}
likelihoods <- mxsure_likelihood(mixed_snp_dist = mixed_data$snp_dist,
  unrelated_snp_dist = distant_data$snp_dist,
  mixed_time_dist = mixed_data$time_dist)
head(likelihoods, 5)
```

Based on these likelihoods we can plot the original data with SNP distance versus time distance for each comparisons, using the `snp_over_time()` function. We colour the points as their likelihoods. We can also plot the predictive intervals and inferred SNP threshold. `snp_over_time()` calls `mxsure_likelihood()` under the hood.

```{r}
snp_over_time(mixed_snp_dist = mixed_data$snp_dist,
  unrelated_snp_dist = distant_data$snp_dist,
  mixed_time_dist = mixed_data$time_dist
  )
```
